umap/distances.py

```diff
@@ -182,6 +182,43 @@ def minkowski_grad(x, y, p=2):
     return result ** (1.0 / p), grad
 
 
+@numba.njit()
+def poincare(u, v):
+    """Poincare distance.
+
+    ..math::
+        \delta (u, v) = 2 \frac{ \lVert  u - v \rVert ^2 }{ ( 1 - \lVert  u \rVert ^2 ) ( 1 - \lVert  v \rVert ^2 ) }
+        D(x, y) = \operatorname{arcosh} (1+\delta (u,v))
+    """
+    sq_u_norm = np.sum(u * u)
+    sq_v_norm = np.sum(v * v)
+    sq_dist = np.sum(np.power(u - v, 2))
+    return np.arccosh(1 + 2 * (sq_dist / ((1 - sq_u_norm) * (1 - sq_v_norm))))
+
+
+@numba.njit()
+def poincare_grad(u, v):
+    """Poincare distance.
+
+    ..math::
+        \delta (u, v) = 2 \frac{ \lVert  u - v \rVert ^2 }{ ( 1 - \lVert  u \rVert ^2 ) ( 1 - \lVert  v \rVert ^2 ) }
+        D(x, y) = \operatorname{arcosh} (1+\delta (u,v))
+    """
+    eps = 1e-6
+    sq_u_norm = np.sum(u * u) - eps
+    sq_v_norm = np.sum(v * v) - eps
+    sq_dist = np.sum(np.power(u - v, 2))
+    result = np.arccosh(1 + 2 * (sq_dist / ((1 - sq_u_norm) * (1 - sq_v_norm))))
+
+    d = np.sqrt(result)
+    e_grad = (u - v) / (eps + d)
+
+    p_sq_norm = 1 - np.sum(np.power([u, v], 2))
+    r_grad = e_grad * (np.power(p_sq_norm, 2) / 4)
```

Comment:
I admit I'm struggling to derive this formulation of the gradient -- could you provide a sketch of the derivation for me?

Reply:
yeah i am still working on this / confirming this is correct, which is mostly why i've left this branch as `[wip]` sorry i can't get to it super soon, but maybe i can find some time late this week

Reply:
It's not a rush -- whenever you get to it. Thanks for working on this!